common/StringVector.hpp

```diff
@@ -111,64 +128,84 @@ class StringVector
     static StringVector tokenize(const char* data, const std::size_t size,
-                                 const char delimiter = ' ')
+                                 const char delimiter = ' ', bool ignoreQuotes = true)
     {
         if (size == 0 || data == nullptr || *data == '\0')
             return StringVector();
 
         std::vector<StringToken> tokens;
-        tokenize(data, size, delimiter, tokens);
+        tokenize(data, size, delimiter, tokens, ignoreQuotes);
         return StringVector(std::string(data, size), std::move(tokens));
     }
 
     /// Tokenize single-char delimited values until we hit new-line or the end.
-    static StringVector tokenize(const std::string& s, const char delimiter = ' ')
+    static StringVector tokenize(const std::string& s, const char delimiter = ' ', bool ignoreQuotes = true)
     {
         if (s.empty())
             return StringVector();
 
         std::vector<StringToken> tokens;
-        tokenize(s.data(), s.size(), delimiter, tokens);
+        tokenize(s.data(), s.size(), delimiter, tokens, ignoreQuotes);
         return StringVector(s, std::move(tokens));
     }
 
     /// Tokenize by the delimiter string.
-    static StringVector tokenize(const std::string& s, const char* delimiter, int len = -1)
+    /// Ignoring delimiters that are within quotation marks, unless ignoreQuotes is set to true
+    static StringVector tokenize(const std::string& s, const char* delimiter, int len = -1, bool ignoreQuotes = true)
     {
         if (s.empty() || len == 0 || delimiter == nullptr || *delimiter == '\0')
             return StringVector();
 
         if (len < 0)
             len = std::strlen(delimiter);
 
-        std::size_t start = 0;
-        std::size_t end = s.find(delimiter, start);
+        std::string delimiterString(delimiter);
+        if(delimiterString.find('\'') != std::string::npos || delimiterString.find('"') != std::string::npos)
+            ignoreQuotes = true;
+
+        // need to track two start indexes:
+        // - one for where we have looked for delimiters(in and out of brackets)
+        // - and another for where to slice the string into tokens
+        std::size_t searchStart = 0;
+        std::size_t tokenStart = 0;
+        std::size_t end = 0;
+
+        int singleQuotesBefore = 0;
+        int doubleQuotesBefore = 0;
 
         std::vector<StringToken> tokens;
         tokens.reserve(16);
 
-        tokens.emplace_back(start, end - start);
-        start = end + len;
-
         while (end != std::string::npos)
         {
-            end = s.find(delimiter, start);
-            tokens.emplace_back(start, end - start);
-            start = end + len;
+            end = s.find(delimiter, searchStart);
+
+            if (end != std::string::npos) {
+                // s.begin() + std::string::npos causes issues
+                singleQuotesBefore = std::count(s.begin(), s.begin() + end, '\'');
```

Comment:
Very costly to do this when `ignoreQuotes` is true. Also, isn't it better to do `find_first_of("\'\"", searchStart, end - searchStart + 1);`

(Actually, the more efficient way to implement support for quotes is to do the search inline. That is, to go through each character and see if it's one of the quotes or the delimiter, and act accordingly. That would at least be a single linear search, where now this algorithm is at least quadratic. But, perhaps it's best to split this into two implementations, the old one that doesn't support quotes, and a new one that might be inefficient but supports quotes.)